File: dsa/linked-lists/go/linked_list.go
package main

import "fmt"

type Node struct {
	Val  int
	Next *Node
}

type LinkedList struct {
	Head *Node
	Size int
}

func NewLinkedList() *LinkedList {
	return &LinkedList{
		Head: nil,
		Size: 0,
	}
}

func (ll *LinkedList) Append(val int) {
	newNode := &Node{Val: val}
	if ll.Head == nil {
		ll.Head = newNode
	} else {
		curr := ll.Head
		for curr.Next != nil {
			curr = curr.Next
		}
		curr.Next = newNode
	}
	ll.Size++
}

func (ll *LinkedList) Prepend(val int) {
	newNode := &Node{Val: val}
	if ll.Head == nil {
		ll.Head = newNode
	} else {
		newNode.Next = ll.Head
		ll.Head = newNode
	}
	ll.Size++
}

func (ll *LinkedList) Pop() {
	if ll.Head == nil {
		return
	}
	if ll.Head.Next == nil {
		ll.Head = nil
		ll.Size--
		return
	}

	curr := ll.Head
	for curr.Next != nil && curr.Next.Next != nil {
		curr = curr.Next
	}
	curr.Next = nil
	ll.Size--
}

func (ll *LinkedList) Remove(val int) {
	if ll.Head == nil {
		return
	}
	curr := ll.Head
	for curr != nil && curr.Next != nil {
		if curr.Next.Val == val {
			curr.Next = curr.Next.Next
			ll.Size--
			return
		}
		curr = curr.Next
	}
}

func (ll *LinkedList) PrintList() {
	if ll.Head == nil {
		return
	} else {
		curr := ll.Head
		for curr != nil {
			fmt.Printf("%d -> ", curr.Val)
			curr = curr.Next
		}
		fmt.Println("nil")
	}
}

func (ll *LinkedList) Reverse() {
	if ll.Head == nil || ll.Head.Next == nil {
		return
	}
	var prev *Node
	curr := ll.Head
	for curr != nil {
		nextNode := curr.Next
		curr.Next = prev
		prev = curr
		curr = nextNode
	}
}

func (ll *LinkedList) GetSize() int {
	return ll.Size
}

func main() {
	fmt.Println("Linked List Implementation")
	fmt.Println("============================")
	fmt.Println("1.Create a linked list")

	ll := NewLinkedList()
	ll.Append(1)
	ll.Append(2)
	ll.Append(3)
	ll.Append(4)
	ll.PrintList()

	fmt.Println("2.Append a value at the end")
	ll.Append(5)
	ll.PrintList()

	fmt.Println("3.Pop a value from the end")
	ll.Pop()
	ll.PrintList()

	fmt.Println("4.Prepend a value at the beginning")
	ll.Prepend(5)
	ll.PrintList()
	fmt.Println("5.Remove a given value")
	ll.Remove(3)
	ll.PrintList()
	fmt.Println("6.Reverse linked list")
	ll.Reverse()
	ll.PrintList()
	fmt.Println("7.Get size of linked list")
	fmt.Println(ll.GetSize())
}

File: dsa/linked-lists/python/linked_list.py

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class LinkedList:
    def __init__(self) -> None:
        self.head = None
        self.size = 0

    def append(self, val) -> None:
        new_node = Node(val)
        if not self.head:
            self.head = new_node
        else:
            curr = self.head
            while curr.next:
                curr = curr.next
            curr.next = new_node
        self.size += 1  

    def prepend(self, val) -> None:
        new_node = Node(val)
        if not self.head:
            self.head = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.size += 1

    def pop(self) -> int | None:
        if not self.head:
            return None

        if not self.head.next:
            val  = self.head.val
            self.head = None
            self.size -= 1
            return val
        curr = self.head
        while curr.next and curr.next.next:
            curr = curr.next
        val = curr.next.val
        curr.next = None
        self.size -= 1
        return val

    def remove(self, val) -> int | None:
        if not self.head:
            return None
        curr = self.head
        while curr.next:
            if curr.next.val == val:
                curr.next = curr.next.next
                self.size -= 1
                return val
            curr = curr.next  
        return None
    
    def get_size(self) -> int:
        return self.size
    
    def print_list(self) -> None:
        curr = self.head
        while curr:
            print(curr.val, end=" -> ")
            curr = curr.next
        print("None")

    def reverse(self) -> None:
        prev = None
        curr = self.head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        self.head = prev

if __name__ == '__main__':
    print("Linked List Implementation")
    print("============================")
    print("1.Create a linked list")

    ll = LinkedList()
    ll.append(1)
    ll.append(2)
    ll.append(3)
    ll.append(4)
    ll.print_list()

    print("2.Append a value at the end")
    ll.append(5)
    ll.print_list()

    print("3.Pop a value from the end")
    val = ll.pop()
    print("Popped value:", val)
    ll.print_list()

    print("4.Prepend a value at the beginning")
    ll.prepend(val)
    ll.print_list()
    print("5.Remove a given value")
    ll.remove(3)
    ll.print_list()
    print("6.Reverse linked list")
    ll.reverse()
    ll.print_list()
    print("7.Get size of linked list")
    print(ll.get_size())
